File: python_cpp/uPIMulator_frontend/src/iss/sram/atomic.py
from typing import List

from abi.word.data_address_word import DataAddressWord
from abi.word.representation import Representation
from iss.sram.lock import Lock
from util.config_loader import ConfigLoader


class Atomic:
    def __init__(self):
        self._address: DataAddressWord = DataAddressWord()
        self._address.set_value(ConfigLoader.atomic_offset())

        self._locks: List[Lock] = [Lock() for _ in range(ConfigLoader.atomic_size())]

    def address(self) -> int:
        return self._address.value(Representation.UNSIGNED)

    def can_acquire(self, address: int) -> bool:
        index = self._index(address)
        return self._locks[index].can_acquire()

    def acquire(self, address: int, id_: int) -> None:
        index = self._index(address)
        self._locks[index].acquire(id_)

    def can_release(self, address: int, id_: int) -> bool:
        index = self._index(address)
        return self._locks[index].can_release(id_)

    def release(self, address: int, id_: int) -> None:
        index = self._index(address)
        self._locks[index].release(id_)

    def _index(self, address: int) -> int:
        index = address - self.address()
        assert 0 <= index < len(self._locks)
        return index

File: python_cpp/uPIMulator_backend/src/simulator/sram/atomic.cc
#include "simulator/sram/atomic.h"

namespace upmem_sim::simulator::sram {

Atomic::Atomic() {
  address_ = new abi::word::DataAddressWord();
  address_->set_value(util::ConfigLoader::atomic_offset());

  size_ = util::ConfigLoader::atomic_size();

  locks_.resize(size_);
  for (int i = 0; i < size_; i++) {
    locks_[i] = new Lock();
  }
}

Atomic::~Atomic() {
  delete address_;

  for (int i = 0; i < size_; i++) {
    delete locks_[i];
  }
}

int Atomic::index(Address address) {
  assert(address >= this->address());
  assert(address < this->address() + size_);
  assert(this->address() <= address and address < this->address() + size_);

  return static_cast<int>(address - this->address());
}

}  // namespace upmem_sim::simulator::sram
